File: engine/dynamic/render_vulkan/src/util/render_pass.cpp
/*
 * This file is a part of Argus.
 * Copyright (c) 2019-2024, Max Roncace <mproncace@protonmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "argus/lowlevel/logging.hpp"

#include "argus/core/engine.hpp"

#include "internal/render_vulkan/defines.hpp"
#include "internal/render_vulkan/setup/device.hpp"
#include "internal/render_vulkan/util/render_pass.hpp"

#include "vulkan/vulkan.h"

namespace argus {
    VkRenderPass create_render_pass(const LogicalDevice &device, VkFormat format, VkImageLayout final_layout,
            bool with_supp_attachments) {
        std::vector<VkAttachmentDescription> atts;
        std::vector<VkAttachmentReference> att_refs;
        atts.reserve(2);
        att_refs.reserve(2);

        VkAttachmentDescription color_att {};
        color_att.format = format;
        color_att.samples = VK_SAMPLE_COUNT_1_BIT;
        color_att.loadOp = VK_ATTACHMENT_LOAD_OP_CLEAR;
        color_att.storeOp = VK_ATTACHMENT_STORE_OP_STORE;
        color_att.stencilLoadOp = VK_ATTACHMENT_LOAD_OP_DONT_CARE;
        color_att.stencilStoreOp = VK_ATTACHMENT_STORE_OP_DONT_CARE;
        color_att.initialLayout = VK_IMAGE_LAYOUT_UNDEFINED;
        color_att.finalLayout = final_layout;
        atts.push_back(color_att);

        VkAttachmentReference color_att_ref {};
        color_att_ref.attachment = SHADER_OUT_COLOR_LOC;
        color_att_ref.layout = VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL;
        att_refs.push_back(color_att_ref);

        if (with_supp_attachments) {
            VkAttachmentDescription light_opac_att {};
            light_opac_att.format = VK_FORMAT_R32_SFLOAT;
            light_opac_att.samples = VK_SAMPLE_COUNT_1_BIT;
            light_opac_att.loadOp = VK_ATTACHMENT_LOAD_OP_CLEAR;
            light_opac_att.storeOp = VK_ATTACHMENT_STORE_OP_STORE;
            light_opac_att.stencilLoadOp = VK_ATTACHMENT_LOAD_OP_DONT_CARE;
            light_opac_att.stencilStoreOp = VK_ATTACHMENT_STORE_OP_DONT_CARE;
            light_opac_att.initialLayout = VK_IMAGE_LAYOUT_UNDEFINED;
            light_opac_att.finalLayout = VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL;
            atts.push_back(light_opac_att);

            VkAttachmentReference light_opac_att_ref {};
            light_opac_att_ref.attachment = SHADER_OUT_LIGHT_OPACITY_LOC;
            light_opac_att_ref.layout = VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL;
            att_refs.push_back(light_opac_att_ref);
        }

        VkSubpassDescription subpass {};
        subpass.pipelineBindPoint = VK_PIPELINE_BIND_POINT_GRAPHICS;
        subpass.colorAttachmentCount = uint32_t(att_refs.size());
        subpass.pColorAttachments = att_refs.data();

        VkSubpassDependency subpass_dep {};
        subpass_dep.srcSubpass = VK_SUBPASS_EXTERNAL;
        subpass_dep.dstSubpass = 0;
        subpass_dep.srcStageMask = VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT;
        subpass_dep.srcAccessMask = 0;
        subpass_dep.dstStageMask = VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT;
        subpass_dep.dstAccessMask = 0;

        VkRenderPassCreateInfo render_pass_info {};
        render_pass_info.sType = VK_STRUCTURE_TYPE_RENDER_PASS_CREATE_INFO;
        render_pass_info.attachmentCount = uint32_t(atts.size());
        render_pass_info.pAttachments = atts.data();
        render_pass_info.subpassCount = 1;
        render_pass_info.pSubpasses = &subpass;
        render_pass_info.dependencyCount = 1;
        render_pass_info.pDependencies = &subpass_dep;

        VkRenderPass render_pass;
        if (vkCreateRenderPass(device.logical_device, &render_pass_info, nullptr, &render_pass) != VK_SUCCESS) {
            crash("Failed to create render pass");
        }

        return render_pass;
    }

    void destroy_render_pass(const LogicalDevice &device, VkRenderPass render_pass) {
        vkDestroyRenderPass(device.logical_device, render_pass, nullptr);
    }
}

File: engine/modules/render_vulkan/src/util/render_pass.rs
use ash::prelude::VkResult;
use ash::vk;
use crate::setup::device::VulkanDevice;
use crate::util::defines::{SHADER_OUT_COLOR_LOC, SHADER_OUT_LIGHT_OPACITY_LOC};

pub(crate) fn create_render_pass(
    device: &VulkanDevice,
    format: vk::Format,
    final_layout: vk::ImageLayout,
    with_supp_attachments: bool
) -> VkResult<vk::RenderPass> {
    let mut atts: Vec<vk::AttachmentDescription> = Vec::with_capacity(2);
    let mut att_refs: Vec<vk::AttachmentReference> = Vec::with_capacity(2);

    let color_att = vk::AttachmentDescription::default()
        .format(format)
        .samples(vk::SampleCountFlags::TYPE_1)
        .load_op(vk::AttachmentLoadOp::CLEAR)
        .store_op(vk::AttachmentStoreOp::STORE)
        .stencil_load_op(vk::AttachmentLoadOp::DONT_CARE)
        .stencil_store_op(vk::AttachmentStoreOp::DONT_CARE)
        .initial_layout(vk::ImageLayout::UNDEFINED)
        .final_layout(final_layout);
    atts.push(color_att);

    let color_att_ref = vk::AttachmentReference::default()
        .attachment(SHADER_OUT_COLOR_LOC)
        .layout(vk::ImageLayout::COLOR_ATTACHMENT_OPTIMAL);
    att_refs.push(color_att_ref);

    if with_supp_attachments {
        let light_opac_att = vk::AttachmentDescription::default()
            .format(vk::Format::R32_SFLOAT)
            .samples(vk::SampleCountFlags::TYPE_1)
            .load_op(vk::AttachmentLoadOp::CLEAR)
            .store_op(vk::AttachmentStoreOp::STORE)
            .stencil_load_op(vk::AttachmentLoadOp::DONT_CARE)
            .stencil_store_op(vk::AttachmentStoreOp::DONT_CARE)
            .initial_layout(vk::ImageLayout::UNDEFINED)
            .final_layout(vk::ImageLayout::SHADER_READ_ONLY_OPTIMAL);
        atts.push(light_opac_att);

        let light_opac_att_ref = vk::AttachmentReference::default()
            .attachment(SHADER_OUT_LIGHT_OPACITY_LOC)
            .layout(vk::ImageLayout::COLOR_ATTACHMENT_OPTIMAL);
        att_refs.push(light_opac_att_ref);
    }

    let subpass = vk::SubpassDescription::default()
        .pipeline_bind_point(vk::PipelineBindPoint::GRAPHICS)
        .color_attachments(att_refs.as_slice());
    let subpasses = [subpass];

    let subpass_dep = vk::SubpassDependency::default()
        .src_subpass(vk::SUBPASS_EXTERNAL)
        .dst_subpass(0)
        .src_stage_mask(vk::PipelineStageFlags::COLOR_ATTACHMENT_OUTPUT)
        .src_access_mask(vk::AccessFlags::empty())
        .dst_stage_mask(vk::PipelineStageFlags::COLOR_ATTACHMENT_OUTPUT)
        .dst_access_mask(vk::AccessFlags::empty());
    let subpass_deps = [subpass_dep];

    let render_pass_info = vk::RenderPassCreateInfo::default()
        .attachments(atts.as_slice())
        .subpasses(&subpasses)
        .dependencies(&subpass_deps);

    let render_pass = unsafe { device.logical_device.create_render_pass(&render_pass_info, None)? };

    Ok(render_pass)
}

pub(crate) fn destroy_render_pass(device: &VulkanDevice, render_pass: vk::RenderPass) {
    unsafe { device.logical_device.destroy_render_pass(render_pass, None) };
}
